README
NAME
    Perinci::Result::Format - Format envelope result

VERSION
    This document describes version 0.38 of Perinci::Result::Format (from
    Perl distribution Perinci-Result-Format), released on 2014-06-18.

SYNOPSIS
DESCRIPTION
    This module formats enveloped result to YAML, JSON, etc. It uses
    Data::Format::Pretty for the backend. It is used by other Perinci
    modules like Perinci::CmdLine and Perinci::Access::HTTP::Server.

    The default supported formats are:

    *   json

        Using Data::Format::Pretty::CompactJSON.

    *   json-pretty

        Using Data::Format::Pretty::JSON.

    *   text-simple

        Using Data::Format::Pretty::SimpleText.

    *   text-pretty

        Using Data::Format::Pretty::Text.

    *   text

        Using Data::Format::Pretty::Console.

    *   yaml

        Using Data::Format::Pretty::YAML.

    *   perl

        Using Data::Format::Pretty::Perl.

    *   phpserialization

        Using Data::Format::Pretty::PHPSerialization.

    *   ruby

        Using Data::Format::Pretty::Ruby.

VARIABLES
%Perinci::Result::Format::Formats => HASH
    Contains a mapping between format names and Data::Format::Pretty::*
    module names + MIME type.

$Enable_Decoration => BOOL (default: 1)
    Decorations include color or other markup, which might make a data
    structure like JSON or YAML string become invalid JSON/YAML. This should
    be turned off if one wants to send the formatting over network.

$Enable_Cleansing => BOOL (default: 0)
    If enabled, cleansing will be done to data to help make sure that data
    does not contain item that cannot be handled by formatter. for example,
    JSON format cannot handle circular references or complex types other
    than hash/array.

FUNCTIONS
    None is currently exported/exportable.

format($res, $format) => STR
    Format enveloped result $res with format named $format.

    Result metadata ("$res->[3]") is also checked for key named
    "result_format_options". The value should be a hash like this "{
    FORMAT_NAME => OPTS, ... }". This way, function results can specify the
    details of formatting. An example enveloped result:

     [200, "OK", ["foo", "bar", "baz"], {
         result_format_options => {
             "text"        => {list_max_columns=>1},
             "text-pretty" => {list_max_columns=>1},
         }
     }]

    The above result specifies that if it is displayed using "text" or
    "text-pretty" format, it should be displayed in one columns instead of
    multicolumns.

RESULT METADATA
    *   property: result_format_options => HASH

FAQ
  How to list supported formats?
    Simply:

     my @supported_formats = keys %Perinci::Result::Format::Formats;

  How to add support for new formats?
    First make sure that Data::Format::Pretty::<FORMAT> module is available
    for your format. Look on CPAN. If it's not, i't also not hard to create
    one.

    Then, add your format to %Perinci::Result::Format::Formats hash:

     use Perinci::Result::Format;

     # this means format named 'xml' will be handled by Data::Format::Pretty::XML
     $Perinci::Result::Format::Formats{xml} = ['XML', 'text/xml'];

SEE ALSO
    Data::Format::Pretty

HOMEPAGE
    Please visit the project's homepage at
    <https://metacpan.org/release/Perinci-Result-Format>.

SOURCE
    Source repository is at
    <https://github.com/sharyanto/perl-Perinci-Result-Format>.

BUGS
    Please report any bugs or feature requests on the bugtracker website
    <https://rt.cpan.org/Public/Dist/Display.html?Name=Perinci-Result-Format
    >

    When submitting a bug or request, please include a test-file or a patch
    to an existing test-file that illustrates the bug or desired feature.

AUTHOR
    Steven Haryanto <stevenharyanto@gmail.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Steven Haryanto.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Perinci-Result-Format

0.38     2014-06-18 (SHARYANTO)

         [BUG FIXES]

         - Second attempt, coding mistake.


0.37     2014-06-18 (SHARYANTO)

         - If $Enable_Cleansing is true, do clone-and-clean instead of
           clean-in-place, to avoid modifying e.g. Rinci metadata cache in a
           long-running server.


0.36     2014-02-14 (SHARYANTO)

         - Update for the perl formatter (it adds color by default now, so we
           need to be able to handle/disable it).


0.35     2013-10-24 (SHARYANTO)

         - Add format: phpserialization.


0.34     2013-10-05 (SHARYANTO)

         - Add formats: perl & ruby.


0.33     2013-10-03 (SHARYANTO)

         - text: Colorize error response.


0.32     2013-09-16 (SHARYANTO)

         - text: Show previous errors and stack trace.


0.31     2013-07-05 (SHARYANTO)

         [ENHANCEMENTS]

         - Add package variable flag $Enable_Cleansing to allow in-place
           cleansing of data that contain items that cannot be handled by
           formatter. For example, JSON cannot handle circular data or complex
           types other than array/hash.


0.30     2013-03-01 (SHARYANTO)

         [ENHANCEMENTS]

         - Add package variable flag $Enable_Decoration to allow disabling
           colored YAML/JSON output (used by Perinci-Access-HTTP-Server, for
           example).


0.29     2012-12-25 (SHARYANTO) - Merry Christmas!

         - Observe result metadata: result_format_options.


0.28     2012-08-10 (SHARYANTO)

         - First release, split to make Perinci dist empty (that dist was a
           growing mixed bag and caused headache in dependencies).


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601",
    "Test::More" => "0.96"
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Format envelope result",
  "dist_author" => [
    "Steven Haryanto <stevenharyanto\@gmail.com>"
  ],
  "dist_name" => "Perinci-Result-Format",
  "dist_version" => "0.31",
  "license" => "perl",
  "module_name" => "Perinci::Result::Format",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Data::Clean::JSON" => 0,
    "Data::Format::Pretty" => 0,
    "Data::Format::Pretty::Console" => "0.12",
    "Data::Format::Pretty::JSON" => "0.03",
    "Data::Format::Pretty::YAML" => "0.02",
    "Rinci" => "v1.1.22",
    "perl" => "5.010001"
  },
  "script_files" => []
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.015.
use strict;
use warnings;

use 5.010001;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Format envelope result",
  "AUTHOR" => "Steven Haryanto <stevenharyanto\@gmail.com>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Perinci-Result-Format",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Perinci::Result::Format",
  "PREREQ_PM" => {
    "Color::ANSI::Util" => 0,
    "Data::Clean::JSON" => "0.15",
    "Data::Dump::OneLine" => 0,
    "Data::Format::Pretty" => 0,
    "Data::Format::Pretty::Console" => "0.12",
    "Data::Format::Pretty::JSON" => "0.10",
    "Data::Format::Pretty::PHPSerialization" => 0,
    "Data::Format::Pretty::Perl" => "0.05",
    "Data::Format::Pretty::Ruby" => 0,
    "Data::Format::Pretty::YAML" => "0.07",
    "Rinci" => "v1.1.38",
    "Term::Detect::Software" => 0
  },
  "TEST_REQUIRES" => {
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => "0.96"
  },
  "VERSION" => "0.38",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Color::ANSI::Util" => 0,
  "Data::Clean::JSON" => "0.15",
  "Data::Dump::OneLine" => 0,
  "Data::Format::Pretty" => 0,
  "Data::Format::Pretty::Console" => "0.12",
  "Data::Format::Pretty::JSON" => "0.10",
  "Data::Format::Pretty::PHPSerialization" => 0,
  "Data::Format::Pretty::Perl" => "0.05",
  "Data::Format::Pretty::Ruby" => 0,
  "Data::Format::Pretty::YAML" => "0.07",
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Rinci" => "v1.1.38",
  "Term::Detect::Software" => 0,
  "Test::More" => "0.96"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Perinci/Result/Format.pm
package Perinci::Result::Format;

use 5.010001;
use strict;
use warnings;

use Scalar::Util qw(reftype);

our $VERSION = '0.38'; # VERSION

our $Enable_Decoration = 1;
our $Enable_Cleansing  = 0;

# text formats are special. since they are more oriented towards human instead
# of machine, we remove envelope when status is 200, so users only see content.

# XXX color theme?

my $format_text = sub {
    my ($format, $res) = @_;

    my $stack_trace_printed;

    my $print_err = sub {
        require Color::ANSI::Util;
        require Term::Detect::Software;

        my $use_color = $ENV{COLOR} // 1;
        my $terminfo = Term::Detect::Software::detect_terminal_cached();
        $use_color = 0 if !$terminfo->{color_depth};
        my $colorize = sub {
            my ($color, $str) = @_;
            if ($use_color) {
                if (ref($color) eq 'ARRAY') {
                    (defined($color->[0]) ?
                         Color::ANSI::Util::ansifg($color->[0]):"").
                               (defined($color->[1]) ?
                                    Color::ANSI::Util::ansibg($color->[1]):"").
                                          $str . "\e[0m";
                } else {
                    Color::ANSI::Util::ansifg($color) . $str . "\e[0m";
                }
            } else {
                $str;
            }
        };

        my $res = shift;
        my $out = $colorize->("cc0000", "ERROR $res->[0]") .
            ($res->[1] ? ": $res->[1]" : "");
        $out =~ s/\n+\z//;
        my $clog; $clog = $res->[3]{logs}[0]
            if $res->[3] && $res->[3]{logs};
        if ($clog->{file} && $clog->{line}) {
            $out .= " (at ".$colorize->('3399cc', $clog->{file}).
                " line ".$colorize->('3399cc', $clog->{line}).")";
        }
        $out .= "\n";
        if ($clog->{stack_trace} && $INC{"Carp/Always.pm"} &&
                !$stack_trace_printed) {
            require Data::Dump::OneLine;
            my $i;
            for my $c (@{ $clog->{stack_trace} }) {
                next unless $i++; # skip first entry
                my $args;
                if (!$c->[4]) {
                    $args = "()";
                } elsif (!ref($c->[4])) {
                    $args = "(...)";
                } else {
                    # periutil 0.37+ stores call arguments in [4]

                    # XXX a flag to let user choose which

                    # dump version
                    #$args = Data::Dump::OneLine::dump1(@{ $c->[4] });
                    #$args = "($args)" if @{$c->[4]} < 2;

                    # stringify version
                    $args = Data::Dump::OneLine::dump1(
                        map {defined($_) ? "$_":$_} @{ $c->[4] });
                    $args = "($args)" if @{$c->[4]} == 1;
                }
                $out .= "    $c->[3]${args} called at $c->[1] line $c->[2]\n";
            }
            $stack_trace_printed++;
        }
        $out;
    };

    if (!defined($res->[2])) {
        my $out = $res->[0] =~ /\A(?:200|304)\z/ ? "" : $print_err->($res);
        my $max = 30;
        my $i = 0;
        my $prev = $res;
        while (1) {
            if ($i > $max) {
                $out .= "  Previous error list too deep, stopping here\n";
                last;
            }
            last unless $prev = $prev->[3]{prev};
            last unless ref($prev) eq 'ARRAY';
            $out .= "  " . $print_err->($prev);
            $i++;
        }
        return $out;
    }
    my ($r, $opts);
    if ($res->[0] == 200) {
        $r = $res->[2];
        my $rfo = $res->[3]{result_format_options} // {};
        # old compat, rfo used to be only opts, now it's {fmt=>opts, ...}
        if ($rfo->{$format}) { $opts = $rfo->{$format} } else { $opts = $rfo }
    } else {
        $r = $res;
        $opts = {};
    }
    if ($format eq 'text') {
        return Data::Format::Pretty::format_pretty(
            $r, {%$opts, module=>'Console'});
    }
    if ($format eq 'text-simple') {
        return Data::Format::Pretty::format_pretty(
            $r, {%$opts, module=>'SimpleText'});
    }
    if ($format eq 'text-pretty') {
        return Data::Format::Pretty::format_pretty(
            $r, {%$opts, module=>'Text'});
    }
};

our %Formats = (
    # YAML::Tiny::Color currently does not support circular refs
    yaml          => ['YAML', 'text/yaml', {circular=>0}],
    json          => ['CompactJSON', 'application/json', {circular=>0}],
    'json-pretty' => ['JSON', 'application/json', {circular=>0}],
    text          => [$format_text, 'text/plain', {circular=>0}],
    'text-simple' => [$format_text, 'text/plain', {circular=>0}],
    'text-pretty' => [$format_text, 'text/plain', {circular=>0}],
    'perl'        => ['Perl', 'text/x-perl', {circular=>1}],
    #'php'         => ['PHP', 'application/x-httpd-php', {circular=>0}],
    'phpserialization' => ['PHPSerialization', 'application/vnd.php.serialized', {circular=>0}],
    'ruby'        => ['Ruby', 'application/x-ruby', {circular=>1}],
);

sub format {
    require Data::Format::Pretty;

    my ($res, $format) = @_;

    my $fmtinfo = $Formats{$format} or return undef;
    my $formatter = $fmtinfo->[0];

    state $cleanser;
    if ($Enable_Cleansing && !$fmtinfo->[2]{circular}) {
        # currently we only have one type of cleansing, oriented towards JSON
        if (!$cleanser) {
            require Data::Clean::JSON;
            $cleanser = Data::Clean::JSON->get_cleanser;
        }
        $res = $cleanser->clone_and_clean($res);
    }

    my $deco = $Enable_Decoration;

    if ((reftype($formatter) // '') eq 'CODE') {
        return $formatter->($format, $res);
    } else {
        my %o;
        $o{color} = 0 if !$deco && $format =~ /json|yaml|perl/;
        return Data::Format::Pretty::format_pretty(
            $res, {%o, module=>$formatter});
    }
}

1;
# ABSTRACT: Format envelope result

__END__

=pod

=encoding UTF-8

=head1 NAME

Perinci::Result::Format - Format envelope result

=head1 VERSION

This document describes version 0.38 of Perinci::Result::Format (from Perl distribution Perinci-Result-Format), released on 2014-06-18.

=head1 SYNOPSIS

=head1 DESCRIPTION

This module formats enveloped result to YAML, JSON, etc. It uses
L<Data::Format::Pretty> for the backend. It is used by other Perinci modules
like L<Perinci::CmdLine> and L<Perinci::Access::HTTP::Server>.

The default supported formats are:

=over 4

=item * json

Using Data::Format::Pretty::CompactJSON.

=item * json-pretty

Using Data::Format::Pretty::JSON.

=item * text-simple

Using Data::Format::Pretty::SimpleText.

=item * text-pretty

Using Data::Format::Pretty::Text.

=item * text

Using Data::Format::Pretty::Console.

=item * yaml

Using Data::Format::Pretty::YAML.

=item * perl

Using Data::Format::Pretty::Perl.

=item * phpserialization

Using Data::Format::Pretty::PHPSerialization.

=item * ruby

Using Data::Format::Pretty::Ruby.

=back

=for Pod::Coverage .*

=head1 VARIABLES

=head1 %Perinci::Result::Format::Formats => HASH

Contains a mapping between format names and Data::Format::Pretty::* module
names + MIME type.

=head1 $Enable_Decoration => BOOL (default: 1)

Decorations include color or other markup, which might make a data structure
like JSON or YAML string become invalid JSON/YAML. This should be turned off if
one wants to send the formatting over network.

=head1 $Enable_Cleansing => BOOL (default: 0)

If enabled, cleansing will be done to data to help make sure that data does not
contain item that cannot be handled by formatter. for example, JSON format
cannot handle circular references or complex types other than hash/array.

=head1 FUNCTIONS

None is currently exported/exportable.

=head1 format($res, $format) => STR

Format enveloped result C<$res> with format named C<$format>.

Result metadata (C<< $res->[3] >>) is also checked for key named
C<result_format_options>. The value should be a hash like this C<< { FORMAT_NAME
=> OPTS, ... } >>. This way, function results can specify the details of
formatting. An example enveloped result:

 [200, "OK", ["foo", "bar", "baz"], {
     result_format_options => {
         "text"        => {list_max_columns=>1},
         "text-pretty" => {list_max_columns=>1},
     }
 }]

The above result specifies that if it is displayed using C<text> or
C<text-pretty> format, it should be displayed in one columns instead of
multicolumns.

=head1 RESULT METADATA

=over

=item * property: result_format_options => HASH

=back

=head1 FAQ

=head2 How to list supported formats?

Simply:

 my @supported_formats = keys %Perinci::Result::Format::Formats;

=head2 How to add support for new formats?

First make sure that Data::Format::Pretty::<FORMAT> module is available for your
format. Look on CPAN. If it's not, i't also not hard to create one.

Then, add your format to %Perinci::Result::Format::Formats hash:

 use Perinci::Result::Format;

 # this means format named 'xml' will be handled by Data::Format::Pretty::XML
 $Perinci::Result::Format::Formats{xml} = ['XML', 'text/xml'];

=head1 SEE ALSO

L<Data::Format::Pretty>

=head1 HOMEPAGE

Please visit the project's homepage at L<https://metacpan.org/release/Perinci-Result-Format>.

=head1 SOURCE

Source repository is at L<https://github.com/sharyanto/perl-Perinci-Result-Format>.

=head1 BUGS

Please report any bugs or feature requests on the bugtracker website L<https://rt.cpan.org/Public/Dist/Display.html?Name=Perinci-Result-Format>

When submitting a bug or request, please include a test-file or a
patch to an existing test-file that illustrates the bug or desired
feature.

=head1 AUTHOR

Steven Haryanto <stevenharyanto@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Steven Haryanto.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
